Inline get bitlen

bjwtaylor opened the pull request:
## Description

Add additional tests to for the get bitlen functionality and simplify the current implementation by in lining it. Resolves https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/522

Verification report provided here 

https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/575#issuecomment-3670594867

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** not required because: No public changes
- [x] **framework PR** not required
- [x] **mbedtls development PR**  not required because: No changes
- [x] **mbedtls 3.6 PR** not required because: No backports
- **tests**  provided

Commit:
Add additional tests for key types and lengths

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
fix handling of public keys

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Inline functions for get_bitlen

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Remove legacy functions and pointers that are now replaced by the inline functions for get_bitlen

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Fix style issues

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Remove definition of get_bitlen function pointer as it is no longer used

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Correct algorithm type for 521 ECC bitlen tests

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Added MBEDTLS_TEST_ALLOW_RSA_4096 to tests that involve 4096 bit rsa keys

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -61,7 +61,7 @@ depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY
 
 PK PSA utilities: RSA setup/free, info functions, unsupported operations 4096 bit
-depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY
+depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY:MBEDTLS_TEST_ALLOW_RSA_4096
```
```suggestion
depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY:PSA_VENDOR_RSA_MAX_KEY_BITS>=4096
```

bjwtaylor replied:
Thanks, I'll add this in

bjwtaylor replied:
I've committed the change, though locally it seems to fail on keys bigger than 2048. Lets see if the CI agrees, though have I missed something?

bjwtaylor replied:
Looks like the CI has failed with that variable set. From the code it looks like maybe I need MBEDTLS_TEST_PK_ALLOW_RSA_KEY_PAIR_4096 or MBEDTLS_TEST_STATIC_KEY_SLOTS_SUPPORT_RSA_4096 as this appears to get set here https://github.com/Mbed-TLS/mbedtls-framework/blob/6c9076eef1aaba371550ebc1145aed7227154b99/tests/include/test/psa_crypto_helpers.h#L534 when the buffer is big enough to store 4096. Though I'm not sure this is what these variables were intended for?

bjwtaylor replied:
I've added a commit that changes the dependency to !MBEDTLS_PSA_STATIC_KEY_SLOTS, which I think it what @davidhorstmann-arm was suggesting. It seems to do the trick in the local tests, though let me know what people think and lets see if the CI agrees.

Commit:
Change MBEDTLS_TEST_ALLOW_RSA_4096 for PSA_VENDOR_RSA_MAX_KEY_BITS>=4096 in rsa 4096 tests

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add 4096 rsa test dependency for PK on  not MBEDTLS_PSA_STATIC_KEY_SLOTS

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

valeriosetti commented on drivers/builtin/src/pk.c:
```diff
@@ -13,6 +13,7 @@
 
 #include "tf_psa_crypto_common.h"
 
+#include "mbedtls/platform.h"
```
I don't see why changes added by this PR require this include. Why is it necessary?

bjwtaylor replied:
It's not required, it should now be removed.

valeriosetti commented on drivers/builtin/src/pk.c:
```diff
@@ -1332,13 +1333,25 @@ int mbedtls_pk_check_pair(const mbedtls_pk_context *pub,
     if (ctx == NULL || ctx->pk_info == NULL) {
         return 0;
     }
+    if (strncmp(ctx->pk_info->name, "Opaque", 6) == 0) {
```
I'm not a huge fan of this approach. Nowadays:
- a PK context with a key pair has `ctx->priv_id` set (EC, RSA, Opaque);
- a PK context with transparent EC|RSA key pair or public key has `ctx->bits` set.

I suggest to simplify this block like this:
```
    if (ctx->bits != 0) {
        return ctx->bits;
    } else {
        if (PSA_SUCCESS != psa_get_key_attributes(ctx->priv_id, &attributes)) {
            return 0;
        }
        bits = psa_get_key_bits(&attributes);
        psa_reset_key_attributes(&attributes);
        return bits;
    }
```

It should work the same (TBH I didn't test it), but it drops the dependency on `strncmp()`.

mpg replied:
Seconded. In addition to Valerio's points, eventually we want to remove `pk_info` entirely, so please don't rely on it in new code. (Also, eventually we want to remove most if not all strings from crypto, so don't rely on string comparison in new code.)

mpg replied:
On second thought, shouldn't we just ensure that `ctx->bits` is always set for all keys, as we do for `psa_type` in #532? I don't really like struct having fields that are sometimes set sometimes not depending on context and you have to remember when - sure it's documented, but how often do we forget to update the documentation when we change the code?

bjwtaylor replied:
So, @valeriosetti comment looks sensible. So I've updated the code to reflect this @mpg I'm not sure what you're after in your final comment. However let me know if you're unhappy with the original suggestion and we can change it.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -619,87 +619,96 @@ exit:
  * END_DEPENDENCIES
  */
 
+/* END_CASE */
```
??

bjwtaylor replied:
Now removed.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -619,87 +619,96 @@ exit:
-    TEST_ASSERT(mbedtls_pk_get_bitlen(&pk) == bitlen);
-    TEST_ASSERT(mbedtls_pk_get_len(&pk) == (bitlen + 7) / 8);
+    TEST_ASSERT(mbedtls_pk_get_bitlen(&priv_pk) == (size_t) bitlen);
+    TEST_ASSERT(mbedtls_pk_get_len(&priv_pk) == ((size_t) bitlen + 7) / 8);
```
```suggestion
    TEST_ASSERT(mbedtls_pk_get_len(&priv_pk) == PSA_BITS_TO_BYTES(bitlen));
```
?

bjwtaylor replied:
Should now be updated.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -619,87 +619,96 @@ exit:
 
+    pub_key = psa_pub_key_from_priv(key);
+
+    TEST_EQUAL(mbedtls_pk_copy_from_psa(pub_key, &pub_pk), 0);
```
I think you can use `mbedtls_pk_copy_public_from_psa()` here and get rid of `psa_pub_key_from_priv()` call and the `pub_key` variable

mpg replied:
I think ideally, here or elsewhere, we want to test `get_bitlen()` both on contexts created with `copy_from_psa()` and with `copy_public_from_psa()` (see my general comment about the test plan). It would probably be convenient to do it here.

However this should either be done using new PK contexts, or after the existing "test that freeing the context does not destroy the key" block, that's currently below.

bjwtaylor replied:
@valeriosetti I've updated this line as per your comment. @mpg I'll include you point on test coverage in a separate commit in response to the test section.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -708,8 +717,11 @@ exit:
+    mbedtls_pk_free(&priv_pk);   /* redundant except upon error */
+    mbedtls_pk_free(&priv_pk2);
+    mbedtls_pk_free(&pub_pk);
+    psa_destroy_key(key);
```
The call to `psa_destroy_key()` here wasn't required before this PR as the key should normally be destroyed just before the `exit` label. Is this really necessary?

mpg replied:
Even if the key is normally destroyed earlier, we should also consider the abnormal path where an assertion fails and we jump to exit early: in that case we still want to destroy the key in order to avoid any leaks. So, I think the previous version was wrong and the new one is more robust.

(Perhaps the comment on line 720 should be moved to the line before to make it clearer it applies to the whole paragraph not just its first line.)

bjwtaylor replied:
Now updated as per @mpg's comment.

mpg commented on tests/suites/test_suite_pk.data:
```diff
@@ -48,13 +48,21 @@ PK utils: ECDSA SECP521R1
+pk_psa_utils:PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):521
+
+PK PSA utilities: RSA setup/free, info functions, unsupported operations 1024 bit
 depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY
```
Pre-existing, but this looks wrong: we'll be creating a key pair, so we should depend on `RSA_KEY_PAIR` not `RSA_PUBLIC_KEY`.

bjwtaylor replied:
Should now be updated.

bjwtaylor replied:
So, if I use this config the tests never get run. I've had a dig around and PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_BASIC seems to be the likely candidate. So I've updated to this, lets see what the CI does and let me know what you think to this choice?

mpg reviewed (changes requested):
In addition to the line comments I left, I'm not convinced by the test plan. There are currently 5 ways to populate a PK context:

1. `mbedtls_pk_wrap_psa()`
2. `mbedtls_pk_copy_from_psa()`
3. `mbedtls_pk_copy_public_from_psa()`
4. `mbedtls_pk_parse_key()`
5. `mbedtls_pk_parse_public_key()`

We want each of them to be appropriately tested (both keypairs and public keys (except for (1)), both RSA and ECC, each with 2 different key sizes).

I'm satisfied that the first case is well covered by the improvements you made to `pk_psa_utils()`.

With minor adaptations (see line comment) this function can probably also cover (2) and (3). On second thought, I'm not sure that would be right: `pk_psa_utils()` is currently focused on wrapping, so adding unrelated stuff just muddies the water. Perhaps having a dedicated testing function would be cleaner. Or re-use `pk_copy_from_psa_success()`?

However I'm more worried about (4) and (5). Superficially, it looks like it should be the job of the `pk_utils()` test function, but it calls `pk_setup()` which constructs the context manually. (Also, it only does keypairs, not public keys.)

So, unless I missed something, (4) and (5) are not covered at all. I'm not sure what's the simplest way to test them at this point. One way would be to add a `bitlen` parameter to the existing parsing functions in `test_suite_pkparse`, but that would probably mean a lot of lines changed. Or we could just add a dedicated function in `test_suite_pk` that takes as input a serialized private or public key (with a parameter indicating which it is), and its expected bitlen, then calls the appropriate parse function, then `get_bitlen()` and we're done.

Wdyt?

valeriosetti commented:
> One way would be to add a bitlen parameter to the existing parsing functions in test_suite_pkparse, but that would probably mean a lot of lines changed. Or we could just add a dedicated function in test_suite_pk that takes as input a serialized private or public key (with a parameter indicating which it is), and its expected bitlen, then calls the appropriate parse function, then get_bitlen() and we're done.

Adding my personal opinion here. Unless changes to already existing test functions is trivial/simple I prefer to create new test functions for specific purposes. Extending what's already there makes test functions hard to read: lot of input parameters, lot of operations carried out internally... IMO this is not easy to review and maintain.

The worst consequence that I can think of in adding new test functions is that we duplicate some test that's already there. This is not bad though: it's true that testing will take slightly longer, but that's better than missing test coverage or make something hard to read.

Commit:
Remove platform.h as it is no longer required

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Remove stray END_CASE

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update to use PSA_BITS_TO_BYTES

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update comments to better describe the implementation

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Improve handling of opaque contexts in mbedtls_pk_get_bitlen

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update pk_psa_utils to use mbedtls_pk_copy_public_from_psa

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update the dependency from RSA_PUBLIC_KEY to RSA_KEY_PAIR for RSA setup/free, info functions, unsupported operations 1024 bit as it is incorrect

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Change minimum RSA size test to 2048

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Updated the 2048 bit rsa test to use PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_BASIC

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

bjwtaylor commented on drivers/builtin/src/pk.c:
```diff
@@ -1341,13 +1341,24 @@ int mbedtls_pk_check_pair(const mbedtls_pk_context *pub,
     }
-
-    return ctx->pk_info->get_bitlen((mbedtls_pk_context *) ctx);
+    if (ctx->bits != 0) {
```
@mpg copying you comment here so we can reply in a thread and it's easier to track:
On second thought, shouldn't we just ensure that ctx->bits is always set for all keys, as we do for psa_type in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/532? I don't really like struct having fields that are sometimes set sometimes not depending on context and you have to remember when - sure it's documented, but how often do we forget to update the documentation when we change the code?

bjwtaylor replied:
Not sure what you're after here? Is it a general architectural question on whether we should be checking somewhere else whether bits != 0 or are you after the above (recently added solution) or something else?

mpg replied:
Thanks for creating a thread :) Yes, that's a general architectural question: once a `pk_context` has been populated (by one of the 5 methods mentioned above), is `ctx->bits` guaranteed to be valid, ie hold the actual bit length of the key?

I think as a general principle, it would be nice for the answer to this question to be "yes", rather than "well, depends which function was used to populate your context - if it was X or Y then yes, otherwise you'll have to do Z instead to get the bitlen". That's a general thought about fields in `pk_context`.

However, for this particular field, since we have a getter function which can be used everywhere (instead of accessing the field), I think it's not that important. So, on second thought, not requesting any change here - especially as this PR is already large enough as it is :)

valeriosetti replied:
Oops, I was about to ask for the very same change that @mpg suggested here when I saw it was already discussed. To me it would make sense as it aligns with what is being done in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/577#discussion_r2597951282.
Since we're going to test `mbedtls_pk_get_bitlen()` with all the possible ways of generating a PK context nowadays, we should be granted that we're not forgetting anything.

I'm deeply sorry to ask for this change since I was the one who proposed the patch for it initially, but the comment I linked above made me realize that this is indeed the right path to follow.
However if we think that this PR is becoming too large, we can address this with a small follow-up PR specifically targeted for this change.

valeriosetti replied:
I'm super sorry, I just realized that I had this unsent message that was pending. I don´t know why it wasn't sent.
I think it's mostly directed to @mpg because the goal is to find a "shared design pattern" in PK that should be used in all the function.

Anyway in case @mpg agrees with my point, the following patch should be enough to make `tests_suite_pk` testing happy again:
```
diff --git a/drivers/builtin/src/pk.c b/drivers/builtin/src/pk.c
index 2f1dfa893..13eb7adab 100644
--- a/drivers/builtin/src/pk.c
+++ b/drivers/builtin/src/pk.c
@@ -163,6 +163,7 @@ int mbedtls_pk_wrap_psa(mbedtls_pk_context *ctx,
     const mbedtls_pk_info_t *info = NULL;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
     psa_key_type_t type;
+    psa_key_bits_t bits;
     int ret;
 
     if (ctx == NULL || ctx->pk_info != NULL) {
@@ -173,6 +174,7 @@ int mbedtls_pk_wrap_psa(mbedtls_pk_context *ctx,
         return MBEDTLS_ERR_PK_BAD_INPUT_DATA;
     }
     type = psa_get_key_type(&attributes);
+    bits = psa_get_key_bits(&attributes);
     psa_reset_key_attributes(&attributes);
 
 #if defined(PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY)
@@ -196,6 +198,7 @@ int mbedtls_pk_wrap_psa(mbedtls_pk_context *ctx,
 
     ctx->pk_info = info;
     ctx->psa_type = type;
+    ctx->bits = bits;
 
     return 0;
 }
@@ -1360,24 +1363,13 @@ int mbedtls_pk_check_pair(const mbedtls_pk_context *pub,
  */
 size_t mbedtls_pk_get_bitlen(const mbedtls_pk_context *ctx)
 {
-    size_t bits;
-    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
-
     /* For backward compatibility, accept NULL or a context that
      * isn't set up yet, and return a fake value that should be safe. */
     if (ctx == NULL || ctx->pk_info == NULL) {
         return 0;
     }
-    if (ctx->bits != 0) {
-        return ctx->bits;
-    } else {
-        if (PSA_SUCCESS != psa_get_key_attributes(ctx->priv_id, &attributes)) {
-            return 0;
-        }
-        bits = psa_get_key_bits(&attributes);
-        psa_reset_key_attributes(&attributes);
-        return bits;
-    }
+
+    return ctx->bits;
 }
 
 /*
```

But please wait for @mpg feedback before trying to integrate this. Meanwhile I will take a look at tests

bjwtaylor replied:
This is a relatively small change, so if we think we need it we can add it to this PR. Not sure if this is what @mpg is also after, though if he agrees this solves the issue we can add this.

mpg replied:
Yes, this patch looks like exactly what I was suggesting, and I still think we should go in this direction. I wouldn't block this PR on it, but the patch looks small enough, so @bjwtaylor if you're still happy to integrate this here, that would also be my preference.

bjwtaylor replied:
So, some of this was already in. Though I've also now added the bits that weren't.

valeriosetti replied:
I'm currently checking the latest version of this PR, but I don't see the changes to `mbedtls_pk_wrap_psa` that I made in my patch above. Weren't they necessary to pass tests?

bjwtaylor replied:
Apologies should now be there.

bjwtaylor replied:
So, I think the bits variable of the patch needs to be of type size_t to prevent a failure in the windows compiler. I've pushed the change let's see what the CI thinks and then we can decide what we think to this change.

valeriosetti replied:
Given the prototype of `psa_get_key_bits` is:
`static size_t psa_get_key_bits(const psa_key_attributes_t *attributes);`
I think your change is fine. Sorry for the mistake in the patch.

bjwtaylor commented on tests/suites/test_suite_pk.function:
```diff
@@ -708,8 +714,12 @@ exit:
+    psa_destroy_key(pub_key);
     PSA_DONE();
 }
 /* END_CASE */
```
@mpg again adding your comments as a thread to collate them better:
In addition to the line comments I left, I'm not convinced by the test plan. There are currently 5 ways to populate a PK context:

1. mbedtls_pk_wrap_psa()
2. mbedtls_pk_copy_from_psa()
3. mbedtls_pk_copy_public_from_psa()
4. mbedtls_pk_parse_key()
5. mbedtls_pk_parse_public_key()

We want each of them to be appropriately tested (both keypairs and public keys (except for (1)), both RSA and ECC, each with 2 different key sizes).

I'm satisfied that the first case is well covered by the improvements you made to pk_psa_utils().

With minor adaptations (see line comment) this function can probably also cover (2) and (3). On second thought, I'm not sure that would be right: pk_psa_utils() is currently focused on wrapping, so adding unrelated stuff just muddies the water. Perhaps having a dedicated testing function would be cleaner. Or re-use pk_copy_from_psa_success()?

However I'm more worried about (4) and (5). Superficially, it looks like it should be the job of the pk_utils() test function, but it calls pk_setup() which constructs the context manually. (Also, it only does keypairs, not public keys.)

So, unless I missed something, (4) and (5) are not covered at all. I'm not sure what's the simplest way to test them at this point. One way would be to add a bitlen parameter to the existing parsing functions in test_suite_pkparse, but that would probably mean a lot of lines changed. Or we could just add a dedicated function in test_suite_pk that takes as input a serialized private or public key (with a parameter indicating which it is), and its expected bitlen, then calls the appropriate parse function, then get_bitlen() and we're done.

Wdyt?

bjwtaylor replied:
@valeriosetti I think your comment is related to the above, so adding it to the same thread:
One way would be to add a bitlen parameter to the existing parsing functions in test_suite_pkparse, but that would probably mean a lot of lines changed. Or we could just add a dedicated function in test_suite_pk that takes as input a serialized private or public key (with a parameter indicating which it is), and its expected bitlen, then calls the appropriate parse function, then get_bitlen() and we're done.

Adding my personal opinion here. Unless changes to already existing test functions is trivial/simple I prefer to create new test functions for specific purposes. Extending what's already there makes test functions hard to read: lot of input parameters, lot of operations carried out internally... IMO this is not easy to review and maintain.

The worst consequence that I can think of in adding new test functions is that we duplicate some test that's already there. This is not bad though: it's true that testing will take slightly longer, but that's better than missing test coverage or make something hard to read.

bjwtaylor replied:
@mpg gave this the thumb up, so I'll add a function that takes a key(s) and calls the function and getbitlen. Let me know. Hopefully this should complete the required test plan, though let me know if anyone thinks we need anything else.

mpg replied:
As discussed, we want to reach a balance between to valid goals: (a) not making tests more difficult to read than they already are, and (b) not stalling PRs looking for perfection.

IMO, here a good solution could be:
- Add testing of `mbedtls_pk_copy_from_psa()` and `mbedtls_pk_copy_public_from_psa()` to the existing function that tests `mbedtls_pk_wrap_psa()` - IMO this is a natural grouping as those are the 3 ways to populate a PK context from a PSA key. This is the approach used in this PR so far, that we can keep and just slightly expand to test both `copy` and `copy_public`.
- OTOH, for keys creating from parsing, don't extend the existing test functions for parsing with a new bitlen parameter (as I previously suggested, but Valerio pointed out this approach is bad in the long run), but instead create a new dedicated testing function "parse and check bitlen". We can probably use the same function for both `parse_key` and `parse_public_key` with a parameter indicating which one it is.

Note that if this works this is probably a general pattern for testing something with a PK context:
- one function for the 3 cases of "populated from a PSA key"
- another function for the 2 cases of "populated by parsing"

Wdyt @bjwtaylor  @valeriosetti ?

mpg replied:
Note: in principle, it is desirable for tests that parse key to parse them from a string in the .data file, as opposed to a file in the filesystem, in order to maintain coverage in configurations without filesystem support (`MBEDTLS_FS_IO` disabled).

However, in the spirit of balance between quality and speed, I'm happy for the parsing tests in this PR (and others in the same series) to use files as inputs, as I think that's both easier to write and to review (not having to document and check long strings of hex).

bjwtaylor replied:
I think both approaches have their pro's and con's, so I don't have a strong preference for either. So if @mpg and @valeriosetti you can agree on one or the other I'm happy with either.

bjwtaylor replied:
Also I've rocket emoji'd the original threads from this, as the discussion has been now moved here.

Commit:
Revert changes to tests

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Revert changes to test data

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

davidhorstmann-arm commented:
I'm sure there is some rework still to go, but at present this PR saves 36B of code size on M55 :slightly_smiling_face: 

| Module                                      | Base   |  PR  | Delta | % Delta |
| --- | --- | --- | --- | --- |
| pk                                        |  3963  |  4015   |  +52  | +1.31% |
| pk_wrap                                |     2090  |  2002   |  -88 |  -4.21% |
| TOTAL                                   |  206690 | 206654  |   -36 |  -0.02% |

Commit:
Add test case for NULL pk_info

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add tests for 2048 bit RSA keys

Key material generated with the following commands:

openssl genrsa -out priv.key 2048
openssl rsa -in priv.key -pubout -out pubkey.key
openssl base64 -d -in pubkey.key -out pubkey.der
openssl base64 -d -in priv.key -out privkey.der
xxd -p pubkey.der | tr -d '\n'
xxd -p privkey.der | tr -d '\n'

and copied into the data file.

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add tests for 1024 bit RSA keys

Key material generated with the following commands:

    openssl genrsa -out priv.key 1024
    openssl rsa -in priv.key -pubout -out pubkey.key
    openssl base64 -d -in pubkey.key -out pubkey.der
    openssl base64 -d -in priv.key -out privkey.der
    xxd -p pubkey.der | tr -d '\n'
    xxd -p privkey.der | tr -d '\n'

 and copied into the data file.

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add Tests for prime256v1 curves

Key material generated with the following commands:

    openssl ecparam -name prime256v1 -genkey -noout -out priv.key
    openssl ec -in priv.key -pubout -out pubkey.key
    openssl base64 -d -in pubkey.key -out pubkey.der
    openssl base64 -d -in priv.key -out privkey.der
    xxd -p pubkey.der | tr -d '\n'
    xxd -p privkey.der | tr -d '\n'

and copied into the data file.

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add Tests for secp521r1 Curves

Key Material Generated with the following commands:

    openssl ecparam -name secp521r1 -genkey -noout -out priv.key
    openssl ec -in priv.key -pubout -out pubkey.key
    openssl base64 -d -in pubkey.key -out pubkey.der
    openssl base64 -d -in priv.key -out privkey.der
    xxd -p privkey.der | tr -d '\n'
    xxd -p pubkey.der | tr -d '\n'

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Tidy up of comments and depends

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Fixed issues with dependencies, destroying resources and style

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add tests for mbedtls_pk_copy_from_psa and mbedtls_pk_copy_public_from_psa

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Remove debug and improve test dependencies

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add dependency and style fixes

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update name of new function to make it more relevant

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

bjwtaylor commented:
So, I believe I've now implemented the requested test plan as per the below verification matrix

<!DOCTYPE html>

<html>
<body>

Test Case | Generating Function | Test Case | Test function | Test case | Line
-- | -- | -- | -- | -- | --
TC1 | non-wrapping -> mbedtls_pk_parse_key | positive: mbedtls_pk_get_bitlen() for EC curve#1 | test_bitlen_setup | Test bitlen of 256 bit ECC keys | [991](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L991)
TC2 | non-wrapping -> mbedtls_pk_parse_key | positive: mbedtls_pk_get_bitlen() for EC curve#2 | test_bitlen_setup | Test bitlen of 521 bit ECC keys | [991](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L991)
TC3 | non-wrapping -> mbedtls_pk_parse_key | positive: mbedtls_pk_get_bitlen() for RSA bit-size#1 | test_bitlen_setup | Test bitlen of 2048 bit RSA keys | [991](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L991)
TC4 | non-wrapping -> mbedtls_pk_parse_key | positive: mbedtls_pk_get_bitlen() for RSA bit-size#2 | test_bitlen_setup | Test bitlen of 1024 bit RSA keys | [991](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L991)
TC6 | non-wrapping -> mbedtls_pk_copy_from_psa | positive: mbedtls_pk_get_bitlen() for EC curve#1 | test_bitlen_setup | Test bitlen of 256 bit ECC keys | [999](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/34573e843f140d691fc39ff297b7729f27d3b4d6/tests/suites/test_suite_pk.function#L999)
TC7 | non-wrapping -> mbedtls_pk_copy_from_psa | positive: mbedtls_pk_get_bitlen() for EC curve#2 | test_bitlen_setup | Test bitlen of 521 bit ECC keys | [999](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L999)
TC8 | non-wrapping -> mbedtls_pk_copy_from_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#1 | test_bitlen_setup | Test bitlen of 2048 bit RSA keys | [999](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L999)
TC9 | non-wrapping -> mbedtls_pk_copy_from_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#2 | test_bitlen_setup | Test bitlen of 1024 bit RSA keys | [999](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L999)
TC10 | non-wrapping -> mbedtls_pk_copy_public_from_psa | positive: mbedtls_pk_get_bitlen() for EC curve#1 | test_bitlen_setup | Test bitlen of 256 bit ECC keys | [1001](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1001)
TC11 | non-wrapping -> mbedtls_pk_copy_public_from_psa | positive: mbedtls_pk_get_bitlen() for EC curve#2 | test_bitlen_setup | Test bitlen of 521 bit ECC keys | [1001](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1001)
TC12 | non-wrapping -> mbedtls_pk_copy_public_from_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#1 | test_bitlen_setup | Test bitlen of 2048 bit RSA keys | [1001](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1001)
TC13 | non-wrapping -> mbedtls_pk_copy_public_from_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#2 | test_bitlen_setup | Test bitlen of 1024 bit RSA keys | [1001](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1001)
TC14 | non-wrapping -> mbedtls_pk_parse_public_key | positive: mbedtls_pk_get_bitlen() for EC curve#1 | test_bitlen_setup | Test bitlen of 256 bit ECC keys | [994](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L994)
TC15 | non-wrapping -> mbedtls_pk_parse_public_key | positive: mbedtls_pk_get_bitlen() for EC curve#2 | test_bitlen_setup | Test bitlen of 521 bit ECC keys |  [994](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L994)
TC16 | non-wrapping -> mbedtls_pk_parse_public_key | positive: mbedtls_pk_get_bitlen() for RSA bit-size#1 | test_bitlen_setup | Test bitlen of 2048 bit RSA keys |  [994](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L994)
TC17 | non-wrapping -> mbedtls_pk_parse_public_key | positive: mbedtls_pk_get_bitlen() for RSA bit-size#2 | test_bitlen_setup | Test bitlen of 1024 bit RSA keys |  [994](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L994)
TC18 | wrapping -> mbedtls_pk_wrap_psa | positive: mbedtls_pk_get_bitlen() for EC curve#1 | test_bitlen_setup | Test bitlen of 256 bit ECC keys | [1009](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1009)
TC19 | wrapping -> mbedtls_pk_wrap_psa | positive: mbedtls_pk_get_bitlen() for EC curve#2 | test_bitlen_setup | Test bitlen of 521 bit ECC keys | [1009](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1009)
TC20 | wrapping -> mbedtls_pk_wrap_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#1 | test_bitlen_setup | Test bitlen of 2048 bit RSA keys | [1009](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1009)
TC21 | wrapping -> mbedtls_pk_wrap_psa | positive: mbedtls_pk_get_bitlen() for RSA bit-size#2 | test_bitlen_setup | Test bitlen of 1024 bit RSA keys | [1009](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/c73ef17c121cfdd3784c4e421dab157058ecd97f/tests/suites/test_suite_pk.function#L1009)
TC22 | wrapping -> mbedtls_pk_wrap_psa | negative: verify that a NULL PK context pointer returns 0 length | valid_parameters |   | [913](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/34573e843f140d691fc39ff297b7729f27d3b4d6/tests/suites/test_suite_pk.function#L913)


</body>

</html>

Let me know what people think.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,51 @@ void valid_parameters()
+
+    mbedtls_pk_init(&pk);
+    mbedtls_pk_init(&pk_pub);
+    mbedtls_pk_init(&new_pk);
```
Invert the order:
```suggestion
    mbedtls_pk_init(&pk);
    mbedtls_pk_init(&pk_pub);
    mbedtls_pk_init(&new_pk);
    
    PSA_INIT();
```

The reason is that `PSA_INIT` can fail and if this happen you are going to free resources with `mbedtls_pk_free` that were never initialized.

bjwtaylor replied:
Good point, should now be corrected.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,51 @@ void valid_parameters()
+
+    TEST_EQUAL(pk_setup(&pk, pk_type, (size_t) curve_or_keybits), 0);
+
+    TEST_ASSERT(mbedtls_pk_parse_key(&pk, priv_key->x, priv_key->len, NULL, 0) == 0);
```
Wait, it seems to me that here you are initializing `pk` twice in a row (`pk_setup` + `mbedtls_pk_parse_key`). I'm a bit surprised that this actually works since the 2nd call should definitely fail because `pk` is already initialized. @mpg what am I missing here?

bjwtaylor replied:
As for why it works, that's one for @mpg. I have however removed the setup though, so at least the duplication isn't present any more and we can hopefully close this one?

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,51 @@ void valid_parameters()
+    TEST_ASSERT(mbedtls_pk_get_bitlen(&pk) == bitlen);
+
+    TEST_ASSERT(mbedtls_pk_parse_public_key(&pk_pub, pub_key->x, pub_key->len) == 0);
+    TEST_ASSERT(mbedtls_pk_get_bitlen(&pk) == bitlen);
```
I think this should be
```suggestion
    TEST_ASSERT(mbedtls_pk_get_bitlen(&pk_pub) == bitlen);
```
because otherwise it's the same as in line 992

bjwtaylor replied:
Yes, good spot this is a copy paste error. Should now be corrected.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,51 @@ void valid_parameters()
+    TEST_ASSERT(mbedtls_pk_parse_public_key(&pk_pub, pub_key->x, pub_key->len) == 0);
+    TEST_ASSERT(mbedtls_pk_get_bitlen(&pk) == bitlen);
+
+    mbedtls_pk_init(&pk_pub);
```
Wait, I think here you re-initializing `pk_pub` (previous initialization is on line 994 with `mbedtls_pk_parse_public_key`) without `mbedtls_pk_free` in between. This is fine for now because `init` is simply going to clear the PK structure and we're not leaking open keys in PSA, but if in the future we want to have PSA key IDs for public keys in PK in the same way we're doing with key pairs then this code would become problematic I think.

bjwtaylor replied:
So I have added the additional free, as I agree this should be required. I guess we also have a general architectural question for @mpg about whether this use of these functions should currently or in the future behave differently?

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,51 @@ void valid_parameters()
+    pub_key_id = psa_pub_key_from_priv(pk.priv_id);
+    TEST_EQUAL(mbedtls_pk_copy_from_psa(pk.priv_id, &new_pk), 0);
+    TEST_ASSERT(mbedtls_pk_get_bitlen(&new_pk) == bitlen);
+    TEST_EQUAL(mbedtls_pk_copy_public_from_psa(pub_key_id, &pk_pub), 0);
```
You don't need `pub_key_id` (i.e. a PSA public key) to call `mbedtls_pk_copy_public_from_psa`. You can use `pk.priv_id` directly and it will work the same. I think you can just remove `pub_key_id` from this function.

bjwtaylor replied:
Thanks for pointing this out, I assumed you would need a separate ID to store the public key. Though I've removed it and everything still works, so I think the change should work.

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -8,6 +8,22 @@ PK write valid parameters
+test_bitlen_setup:"308204bd020100300d06092a864886f70d0101010500048204a7308204a30201000282010100acbe539ee9bc77ec22828db2d0f2786b37f6f5484d14c3383efb0214dd4387b834a1d234c4f64dfa8ad6d8a434571cec60e3dfe0fc1ccc12689030e7b89ed0908f2a6516ef20eb55d1f9089f61813cb46b80600febc6312e38ad5695682cf06df4528b3013ec78d05940fadf82d655ca37604dffbffb810a169e316f770e273d8986918a72d3da56b99ddcfd49d612f016e397f62dca7ff7 ...
+
+Test bitlen of 256 bit ECC keys
+depends_on:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_ECDSA
```
Since we're not going to perform any operation why do we need to ask for ECDSA? Isn't this enough:
```suggestion
depends_on:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT
```
The same holds also for the other EC test below

bjwtaylor replied:
Should now be updated.

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -8,6 +8,22 @@ PK write valid parameters
 valid_parameters_pkwrite:"308204a20201000282010100a9021f3d406ad555538bfd36ee82652e15615e89bfb8e84590dbee881652d3f143504796125964876bfd2be046f973beddcf92e1915bed66a06f8929794580d0836ad54143775f397c09044782b0573970eda3ec15191ea8330847c10542a9fd4cc3b4dfdd061f4d1051406773130f40f86d81255f0ab153c6307e1539acf95aee7f929ea6055be7139785b52392d9d42406d50925897507dda61a8f3f0919bead652c64eb959bdcfe415e17a6da6 ...
 
+Test bitlen of 1024 bit RSA keys
+depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY:MBEDTLS_PKCS5_C
```
Since we're importing a key pair I think that 
```suggestion
depends_on:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_IMPORT:MBEDTLS_PKCS5_C
```
would be better. Wdyt?
The same holds also for the other RSA test below

bjwtaylor replied:
Agree, should now be updated.

valeriosetti reviewed (changes requested):
The newly introduced test function is pretty compact, but it covers most of the cases. The only missing one is the PK context generated with `mbedtls_pk_wrap_psa` for which you linked tests to be [here](https://github.com/bjwtaylor/TF-PSA-Crypto/blob/34573e843f140d691fc39ff297b7729f27d3b4d6/tests/suites/test_suite_pk.function#L1093) but it seems to me that that function is testing `check_pair` and not `get_bitlen`. Am I missing something?

Commit:
Switch the order of init's so that mbedtls init is last to prevent free of resource that don't exist

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Removed duplicate setup as it's not required

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Correct copy paste error in bitlen tests

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add additional free for re-initialisation of pk_pub, to ensure underlying structures are correctly released

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Remove pub_key_id as the public key can be referenced from pk.priv_id

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Replace PSA_WANT_ALG_ECDSA with PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT, as this symbol better represents the dependency

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Update the dependencies of the RSA bitlen tests to better reflect what is used

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add wrap bit tests

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

bjwtaylor commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,59 @@ void valid_parameters()
+    TEST_EQUAL(mbedtls_pk_import_into_psa(&pk, &wrapped_key_attr, &wrapped_key_id), 0);
+    mbedtls_pk_free(&pk);
+    mbedtls_pk_init(&pk);
+    TEST_EQUAL(mbedtls_pk_wrap_psa(&pk, wrapped_key_id), 0);
```
@valeriosetti, good spot on the wrap. I'm not sure how it got missed, probably in the pre-christmas rush to get something out before everyone went on holiday. I've added the test here now though. I'll close the previous comment and update the verification matrix. Then lets see what the CI thinks on this change.

bjwtaylor replied:
Looks like the CI is happy, so hopefully this is good to go.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,59 @@ void valid_parameters()
+    size_t bitlen;
+
+    mbedtls_svc_key_id_t wrapped_key_id = MBEDTLS_SVC_KEY_ID_INIT;
+    psa_key_attributes_t wrapped_key_attr = PSA_KEY_ATTRIBUTES_INIT;
+
+    if ((pk_type == MBEDTLS_PK_RSA) || (pk_type == MBEDTLS_PK_RSASSA_PSS)) {
+        bitlen = curve_or_keybits;
+    } else if (pk_type == MBEDTLS_PK_ECDSA) {
+        TEST_ASSERT(mbedtls_ecc_group_to_psa(curve_or_keybits, &bitlen) != 0);
+    } else {
+        TEST_FAIL("Key type not supported");
+    }
```
I think you can simplify this now: `pk_type` and `curve_or_keybits` are overkilling, you just need to know the bit size to use as check below. So why not just using that as input parameter and getting rid of this block?

bjwtaylor replied:
So the aim of the current architecture is to move the intelligence of how the  keybits is calculated into the function, to reduce maintenance. Yes we could add another parameter to the function, though this would then mean every time we need to add a new call in the data file you would need to add this to it. We'd also need to change each one individually if we needed to refactor anything. So I think the extra complexity is justified, though happy to change it if you feel strongly that it's not worth it. What do you think?

valeriosetti replied:
Actually my proposal was to remove `pk_type` and `curve_or_keybits` and replace them with `bitsize` parameter, so it's not adding a new parameter, but reducing them. In any case, whenever you add a new line to the data file using this function, you need to specify either the bitsize (my solution) or `curve_or_keybits` (your solution), because in both cases you need to have a "known good value" for comparison.

Anyway I don´t have a strong preference, both solution work just fine IMO, so feel free to ignore this comment ;)

valeriosetti reviewed (changes requested):
Mostly look good to me. I only left one possible improvement to the test function.
There is however one [unresolved comment](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/575#discussion_r2635276687) that is not addressed and I think it needs @mpg feedback.

Commit:
Refactor get_bitlen to remove legacy functionality

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Add handling of bits to pk_wrap_psa

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

valeriosetti reviewed (dismissed):
LGTM. Thanks!

Commit:
Update framework to fix crypto-common.make failure

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Converting the variable the holds the output of psa_get_key_bits to size_t, as it causes issues with the windows compiler

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

valeriosetti reviewed (dismissed):
The last commit [looks fine](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/575#discussion_r2681776743), so I'm re-approving

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,59 @@ void valid_parameters()
 }
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_PK_PARSE_C:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC */
```
This function looks like it is common to RSA and ECC, should it really depend on `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC`? I think the whole function shouldn't, on only the ECC-specific parts within the function should be guarded with `#if defined(...)`. (And the specific test cases that use ECC need the appropriate dependencies, but that's already the case.)

bjwtaylor replied:
I've updated this to remove the dependency on PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC, from what I can tell the dependencies on the individual test cases should be sufficient to cover it. Let me know what you think and we will see if the CI agrees.

mpg replied:
I strongly suspect those two lines
```
    } else if (pk_type == MBEDTLS_PK_ECDSA) {
        TEST_ASSERT(mbedtls_ecc_group_to_psa(curve_or_keybits, &bitlen) != 0);
```
need to be guarded with something to ensure `mbedtls_ecc_group_to_psa()` exists in this build. That might be why `depends.py` failed (haven't checked yet).

mpg replied:
Bingo! In `depends.py !PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC` when compiling this file:
```
[2026-01-13T20:13:28.011Z] suites/test_suite_pk.function:984:21: error: implicit declaration of function 'mbedtls_ecc_group_to_psa' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
[2026-01-13T20:13:28.011Z]         TEST_ASSERT(mbedtls_ecc_group_to_psa(curve_or_keybits, &bitlen) != 0);
[2026-01-13T20:13:28.011Z]                     ^
```
This function's declaration is guarded by `#if defined(PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY)` so we should use the same guard when using it.

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -964,6 +965,59 @@ void valid_parameters()
+        TEST_FAIL("Key type not supported");
+    }
+
+    mbedtls_pk_init(&pk);
```
All calls to `_init()` that have a corresponding `_free()` call in the exit block need to go before the first possible failure. Otherwise we might jump to `exit:` and call `_free()` on uninitialised data, which may result in all manner of unpleasant outcomes. It's not a huge deal since it's test code, but we should still demonstrate good practices.

[rant]I've been advocating for years that every structure should come with an `_INIT` macro so that we can always do `mbedtls_pk_context pk = MBEDTLS_PK_INIT;` and not have to think. We keep placing the `_init` call too late in test code, and if we make this mistake in test code, then it's likely some users are making the same mistake in places where it matters much more than test code. IMO initialisation during declaration is a much more robust pattern, and  I think most PSA types allows it, so only the legacy types from PK, X.509 and TLS are missing it. Guess it's just never really the time to do it...[/rant]

bjwtaylor replied:
I've moved them above the first test, hopefully this resolves the issue. Though let me know if you were after something else.

mpg reviewed (changes requested):
Looks pretty good to me, just two nits in the test function.

I like how tests are structured. In the data file it's easy to see which key types & sizes are tested, and in the test function it's easy to see we're testing all 5 methods of populating a context. This makes the (positive part of the) test matrix both complete and, perhaps more importantly easy to understand and extend in the future (eg when we add ML-DSA support). Thank you for that.

Just one point though: is there any place where we're testing that `mbedtls_pk_get_bitlen(&pk)` on a context that was just initialised (`pk_init()`) but never populated successfully returns 0? I couldn't find it. If there is non, I suggest adding one in `valid_parameters()` (just call `pk_get_bitlen` right after the call to `pk_init()`).

bjwtaylor commented:
@mpg in response to your question on the freshly initialised context bitlen=0, if I've understood what you;re after correctly we do that here https://github.com/bjwtaylor/TF-PSA-Crypto/blob/acde1fd41679f121bd55264cb20b4d5a1c61053f/tests/suites/test_suite_pk.function#L914

Commit:
Remove test function dependency on PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC as this is covered by the test data

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

Commit:
Move init functions above tests to prevent a free on an unitialised context

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

mpg reviewed (commented):
At e44f2ed5cc7e6c6fb1b797c0bdb834f6dc54ee21 LGTM except for the CI failure in `depends.py`.

I checked and the other CI failure, in the memsan component, looks like a flaky test and unrelated to this PR, so it can be ignored (unless/until it repeats).

Commit:
Improve config checks for PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC, prior to using dependent functions

Signed-off-by: Ben Taylor <ben.taylor@linaro.org>

mpg commented:
> @mpg in response to your question on the freshly initialised context bitlen=0, if I've understood what you;re after correctly we do that here https://github.com/bjwtaylor/TF-PSA-Crypto/blob/acde1fd41679f121bd55264cb20b4d5a1c61053f/tests/suites/test_suite_pk.function#L914

Ah, I looked there and I missed it as I my attention was drawn to the line above where we pass NULL. Yep, you've understood what I'm after correctly, I had just missed it :)

mpg reviewed (approved):
At 7d20c6b3520a09e2ca108b7800e4ece25ac583a4 LGTM, hoping the CI agrees.
